### PR TITLE
Buckets optimization

### PR DESCRIPTION
**Pull request by grandizzy**
This PR is for reducing contract size, cleaning out the Buckets library usage and preparing for merging `removeAllCollateral` with `removeCollateral` code
- make sure `Bucket` struct is loaded only once from storage and passed to library where is needed (previously functions like `getExchangeRate` was loading it regardless so that resulted in loading bucket several times in the same function call)
- `_isCollateralized` function added to calculate and compare result instead  old`_collateralization` that was just calculating
- kick: load borrower in storage and save inline, mutate `kickPenalty` to t0, check if borrower kickable before doinf expensive operations like pool debt accrue
- accrue pool interest only after checks are done (where is possible) to avoid paying higher gas on reverts

```
============ Deployment Bytecode Sizes ============
  ERC721Pool               -  23,883B  (97.18%)
  ERC20Pool                -  23,877B  (97.15%)
```

vs

```
============ Deployment Bytecode Sizes ============
  ERC20Pool                -  24,098B  (98.05%)
  ERC721Pool               -  24,045B  (97.84%)
```
Gas comparison (shows higher for takes and addQuoteToken which I don't have a good explanation for, but lower for the others which is expected)

```
│ addQuoteToken                              ┆ 99628           ┆ 162400 ┆ 141478 ┆ 668973  ┆ 48001   │
│ arbTake                                    ┆ 426199          ┆ 426289 ┆ 426289 ┆ 426379  ┆ 2       │
│ borrow                                     ┆ 166228          ┆ 228036 ┆ 191545 ┆ 799278  ┆ 104002  │
│ kick                                       ┆ 171327          ┆ 249924 ┆ 244623 ┆ 1091439 ┆ 32000   │
│ moveQuoteToken                             ┆ 286722          ┆ 286722 ┆ 286722 ┆ 286722  ┆ 1       │
│ pledgeCollateral                           ┆ 62338           ┆ 64765  ┆ 64559  ┆ 539664  ┆ 96001   │
│ removeQuoteToken                           ┆ 158409          ┆ 176787 ┆ 183942 ┆ 208941  ┆ 4000    │
│ repay                                      ┆ 514080          ┆ 588688 ┆ 600598 ┆ 840177  ┆ 16001   │
│ take                                       ┆ 53199           ┆ 56839  ┆ 56220  ┆ 278006  ┆ 15998   │
```
vs develop

```
│ addQuoteToken                              ┆ 91419           ┆ 154831 ┆ 133004 ┆ 669816  ┆ 48001   │
│ arbTake                                    ┆ 423267          ┆ 423357 ┆ 423357 ┆ 423447  ┆ 2       │
│ borrow                                     ┆ 166192          ┆ 228063 ┆ 191509 ┆ 800057  ┆ 104002  │
│ kick                                       ┆ 173491          ┆ 252090 ┆ 246789 ┆ 1098650 ┆ 32000   │
│ moveQuoteToken                             ┆ 287572          ┆ 287572 ┆ 287572 ┆ 287572  ┆ 1       │
│ pledgeCollateral                           ┆ 62434           ┆ 64861  ┆ 64655  ┆ 540575  ┆ 96001   │
│ removeQuoteToken                           ┆ 159322          ┆ 177700 ┆ 184855 ┆ 209853  ┆ 4000    │
│ repay                                      ┆ 514982          ┆ 589539 ┆ 601398 ┆ 840977  ┆ 16001   │
│ take                                       ┆ 51506           ┆ 55309  ┆ 54689  ┆ 274876  ┆ 15998   │
```

### CODE REVIEW

File: src/erc721/ERC721Pool.sol [hunk 1]
```diff
@@ -179,15 +179,21 @@ contract ERC721Pool is IERC721Pool, Pool {
      *  @param price_      Price to calculate collateralization for.
      *  @return Collateralization value.
      */
-    function _collateralization(
+    function _isCollateralized(
```
Comment (MikeHathaway) [inline src/erc721/ERC721Pool.sol:182]: Natspec could be updated to be slightly clearer about being a boolean state check.

Also, is there still a need for external clients to be able to view collateralization levels as currently understood by the pool contract?

File: src/erc721/ERC721Pool.sol [hunk 2]
```diff
@@ -179,15 +179,21 @@ contract ERC721Pool is IERC721Pool, Pool {
...
         uint256 debt_,
         uint256 collateral_,
         uint256 price_
-    ) internal pure override returns (uint256) {
-        uint256 encumbered = price_ != 0 && debt_ != 0 ? Maths.wdiv(debt_, price_) : 0;
+    ) internal pure override returns (bool) {
+        if (debt_  == 0) return true;       // if debt is 0 then is collateralized
+        if (price_ == 0) return true;       // if price to calculate collateralized is 0 then is collateralized
+
+        uint256 encumbered = Maths.wdiv(debt_, price_); // calculated to avoid situation where debt dust amount
```
Comment (MikeHathaway) [inline src/erc721/ERC721Pool.sol:190]: Under what circumstances would encumbered ever be 0? Can this be removed?

Also, line 191 has an alignment issue.

File: src/base/Pool.sol [hunk 3]
```diff
@@ -174,13 +189,15 @@ abstract contract Pool is Clone, Multicall, IPool {
         else {
             redeemedLPs_ = Maths.min(lenderLPsBalance, Maths.wrdivr(removedAmount_, exchangeRate));
         }
-        deposits.remove(index_, removedAmount_);  // update FenwickTree
+        deposits.remove(index_, removedAmount_);  // remove from deposits
```
Comment (EdNoepel) [inline src/base/Pool.sol]: The old comment was more useful.  Kind of like seeing `++i;  // increment i`.

File: src/libraries/Buckets.sol [hunk 4]
```diff
@@ -207,41 +179,33 @@ library Buckets {
...
-        uint256 bucketLPs        = self[index_].lps;
-        if (bucketLPs == 0) return (Maths.RAY, bucketCollateral);
-
-        uint256 bucketSize = quoteToken_ * 10**18 + PoolUtils.indexToPrice(index_) * bucketCollateral;  // 10^36 + // 10^36
-        return (bucketSize * 10**18 / bucketLPs, bucketCollateral); // 10^27
+        Bucket storage bucket_,
+        uint256 bucketDeposit_,
+        uint256 bucketPrice_
+    ) internal view returns (uint256) {
+        return bucket_.lps == 0 ? Maths.RAY :
```
Comment (MikeHathaway) [inline src/libraries/Buckets.sol:207]: Would be helpful to keep the precision comments

File: src/base/Pool.sol [hunk 5]
```diff
@@ -641,39 +672,40 @@ abstract contract Pool is Clone, Multicall, IPool {
      *  @param price_      Price to calculate collateralization for.
      *  @return Collateralization value.
```
Comment (EdNoepel) [inline src/base/Pool.sol]: Recommend improving natspec for the return value.

File: src/base/Pool.sol [hunk 6]
```diff
@@ -641,39 +672,40 @@ abstract contract Pool is Clone, Multicall, IPool {
...
         uint256 price_
-    ) internal virtual returns (uint256) {
-        uint256 encumbered = price_ != 0 && debt_ != 0 ? Maths.wdiv(debt_, price_) : 0;
-        return encumbered != 0 ? Maths.wdiv(collateral_, encumbered) : Maths.WAD;
+    ) internal virtual returns (bool) {
+        if (debt_  == 0) return true;       // if debt is 0 then is collateralized
+        if (price_ == 0) return true;       // if price to calculate collateralized is 0 then is collateralized
+
+        uint256 encumbered = Maths.wdiv(debt_, price_); // calculated to avoid situation where debt dust amount
+        if (encumbered   == 0) return true;  // if encumbered is 0 then is collateralized
```
Comment (EdNoepel) [inline src/base/Pool.sol]: nit: alignment

File: src/base/Pool.sol [hunk 7]
```diff
@@ -899,23 +931,26 @@ abstract contract Pool is Clone, Multicall, IPool {
...
             deposit_,
             lpTokens_,
             deposit_,
-            index_
+            PoolUtils.indexToPrice(index_)
         );
     }
 
+    // TODO: remove this method after solving // FIXME: now getting InsufficientLPs with this amount; was working two days ago
     function lpsToCollateral(
```
Comment (MikeHathaway) [inline src/base/Pool.sol:943]: Is this an issue with the ordering of `accruePoolInterest` having differing deposit amounts then before?

File: src/erc20/ERC20Pool.sol [hunk 8]
```diff
@@ -87,25 +87,35 @@ contract ERC20Pool is IERC20Pool, Pool {
     ) external override returns (uint256 fromBucketLPs_, uint256 toBucketLPs_) {
         if (fromIndex_ == toIndex_) revert MoveToSamePrice();
 
-        PoolState memory poolState = _accruePoolInterest();
```
Comment (MikeHathaway) [inline src/erc20/ERC20Pool.sol:90]: Are the gas savings to the revert path a worthwhile consideration here, vs potential accuracy

File: src/base/Pool.sol [hunk 9]
```diff
@@ -641,39 +672,40 @@ abstract contract Pool is Clone, Multicall, IPool {
...
 
-            if (
-                _collateralization(
-                    poolState_.accruedDebt,
-                    poolState_.collateral,
-                    lup_
-                ) != Maths.WAD
-            ) {
-
+            if (poolState_.accruedDebt != 0) {
```
Comment (EdNoepel) [inline src/base/Pool.sol:707]: Good improvement.

File: src/erc20/ERC20Pool.sol [hunk 10]
```diff
@@ -87,25 +87,35 @@ contract ERC20Pool is IERC20Pool, Pool {
...
+            fromBucketLPs_
+        );
+        toBucketLPs_ = Buckets.addCollateral(
+            buckets[toIndex_],
+            deposits.valueAt(toIndex_),
+            collateralAmountToMove_,
+            PoolUtils.indexToPrice(toIndex_)
+        );
 
+        PoolState memory poolState = _accruePoolInterest();
```
Comment (EdNoepel) [inline src/erc20/ERC20Pool.sol]: That's cheating; interest should accrue before LP is moved.  If there is deposit in the bucket, and the bucket is above the HTP, LP holders would be entitled to a slightly larger amount of collateral.

**Review by EdNoepel — CHANGES_REQUESTED** [review]

File: src/erc20/ERC20Pool.sol (hunk 10 above)
Comment (grandizzy) [reply src/erc20/ERC20Pool.sol]: ah, good point, totally forgot that `_accruePoolInterest` is also touching the deposits, will change

File: src/erc721/ERC721Pool.sol (hunk 1 above)
Comment (grandizzy) [reply src/erc721/ERC721Pool.sol:182]: :+1: will update natspec. there is now a `PoolInfoUtils` contract that can be used to pool info https://github.com/ajna-finance/contracts/blob/buckets-optimization/src/base/PoolInfoUtils.sol

File: src/erc721/ERC721Pool.sol (hunk 2 above)
Comment (grandizzy) [reply src/erc721/ERC721Pool.sol:190]: yeah, hit that couple of times too :) encumbered can be 0 if debt is really low, like dust amount, so when divided with price will result in 0

File: src/base/Pool.sol (hunk 7 above)
Comment (grandizzy) [reply src/base/Pool.sol:943]: the issue was introduced with https://github.com/ajna-finance/contracts/commit/210fec425c1aefae37f28fc5915455105acd8b99 that removed `removeAllQuoteTokens` and merged it with `removeQuoteTokens` where there's a dust amount remaining

File: src/erc20/ERC20Pool.sol (hunk 8 above)
Comment (grandizzy) [reply src/erc20/ERC20Pool.sol:90]: I think you're right, I will revert such and keep accuracy (see also this that I totally missed https://github.com/ajna-finance/contracts/pull/325/files/50b450e5d8a9d5672d8efcab102d889bcc7f84cc#r1018243599), will revert that commit

File: src/base/Pool.sol (hunk 3 above)
Comment (grandizzy) [reply src/base/Pool.sol]: reverted in https://github.com/ajna-finance/contracts/pull/325/commits/057173af86ea5b1742f9caaf5d0f18394342931c

File: src/base/Pool.sol (hunk 5 above)
Comment (grandizzy) [reply src/base/Pool.sol]: added with https://github.com/ajna-finance/contracts/pull/325/commits/057173af86ea5b1742f9caaf5d0f18394342931c

**Review by EdNoepel — APPROVED** [review]
LGTM

**Review by MikeHathaway — APPROVED** [review]
LGTM

**Pull request merged**